Documentation about custom camera calibrations (#33)
annotell/public-docs: docs-src/docs/kognic-io/calibrations/cameras-custom.md

---
title: Custom Camera Calibrations
---

If your calibration is not in the list of supported standard camera calibrations, you can provide a custom calibration. The table below
show the attributes of the `CustomCameraCalibration` object. The `wasm_base64` attribute is a base64 encoded string of the
WebAssembly module that implements the calibration. The `test_cases` attribute is a list of `TestCase` objects that are used

> to validate the calibration when during creation. It is recommended to provide a few test cases to make sure that the

Reviewer: ```suggestion
to validate the calibration during creation. It is recommended to provide a few test cases to make sure that the
```

calibration is correct. The `CustomCameraCalibration` object can conveniently be instantiated directly from the binary or 
a wasm-file:
```python
calibration_file = CustomCameraCalibration.from_file("/path/to/calibration.wasm", ...)
calibration_binary = CustomCameraCalibration.from_bytes(b"...", ...)
```

We provide a set of utilities that will make it easier to work with the WebAssembly module. More specifically we provide
validation code and compilation code from a few different languages to WebAssembly. The code is available both as python
functions and via the `kognicutil` cli.

> | Key                   | Value                         | Parameters         |

Reviewer: Move this table to come directly after "The table below
show the attributes of the `CustomCameraCalibration` object."

|:----------------------|:------------------------------|:-------------------|
| `rotation_quaternion` | A `RotationQuaternion` object | `w`, `x`, `y`, `z` |
| `position`            | A `Position` object           | `x`, `y`, `z`      |
| `image_width`         | Integer                       | NA                 |
| `image_height`        | Integer                       | NA                 |
| `wasm_base64`         | String                        | NA                 |
| `test_cases`          | A list of `TestCase` objects  | point3d, point2d   |

## The WebAssembly module

The WebAssembly must follow a strict interface where the module exports a function called `project_point_to_image`. The

> function must take three arguments of type `float64` and return two values of type `float64`. So the text representation

Reviewer: ```suggestion
function must take three arguments of type `float64` and return two values of type `float64`. Thus, the WebAssembly text representation of
```

this interface is `func (param f64 f64 f64) (result f64 f64)`. The three arguments are the x, y and z coordinates of 
the 3D point. The two return values are the x and y coordinates of the projected point in the image plane. WebAssembly
doesn't support multiple return values by default but this can be enabled with the [multi-value](https://github.com/WebAssembly/multi-value)
proposal.

If the point is not within the field of view, the function should return [NaN](https://en.wikipedia.org/wiki/NaN) for 
both the x and y coordinates.

## Validation

:::note

> This requires `wasmtime` to be installed which is an optional dependency kognic-io. Run `pip install kognic-io[wasm]` to

Reviewer: ```suggestion
This requires `wasmtime` to be installed which is an optional of dependency kognic-io. Run `pip install kognic-io[wasm]` to
```

install it.
:::

> We provide validation code both as python functions and via the `kognicutil` cli. We validate things such as but limited to

Reviewer: ```suggestion
We provide validation code both as python functions and via the `kognicutil` cli. We validate things such as but not limited to
```

```

## Compilation

:::note
It is recommended to keep the wasm file as small as possible. Try to avoid dependencies that are not needed. For example,
it may be preferred to implement some mathematical functions yourself instead of using the standard library.
:::

As stated above the WebAssembly module must follow a strict interface and compilation requires the multi-value proposal.
We provide a set of utilities that will make it easier to compile the WebAssembly module from a few languages, see table below. 
The utilities are available both as python functions and via the `kognicutil` cli. From python you can compile the module
with 

```python
from kognic.io.tools.calibration.compilation import compile_to_wasm
wasm_binary = compile_to_wasm("path/to/source")
```

The returned binary can then be used to create a `CustomCameraCalibration` object. If the `output_wasm` parameter is passed,
the binary will be saved to the specified path. The `kognicutil` cli can be used as follows
```bash
koignicutil wasm compile path/to/source path/to/output.wasm
```

Note that, validation is run by default after compilation. This can be disabled with the `--skip-validation` flag.

> | **Language** | **Target**  | **Compilation tool** |

Reviewer: Put the table directly after "languages, see table below."

|--------------|-------------|----------------------|
| Rust         | *.rs        | rustc                |
| Rust (Cargo) | Cargo.toml  | cargo                |
| C++          | *.cc, *.cpp | emscripten           |
| C            | *.c         | emscripten           |

Calibration parameters have to be embedded in the binary so that they can be used by the WebAssembly module. Try to 
pre-compute as much as possible to increase the speed of the projection function at runtime.

Below follows examples of a simplified version of the pinhole calibration in a few different languages.

### Example: Rust

> A [Rust file](https://github.com/annotell/kognic-io-python/blob/1aa5a12496296ec943d68ec94eefab42158fedac/examples/calibration/custom/pinhole.rs)

Author: Change urls here when the PR in kognic-io is merged
